sklearnex/spmd/decomposition/tests/test_incremental_pca_spmd.py

```diff
@@ -0,0 +1,266 @@
+# ==============================================================================
+# Copyright 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose
+
+from onedal.tests.utils._dataframes_support import (
+    _as_numpy,
+    _convert_to_dataframe,
+    get_dataframes_and_queues,
+)
+from sklearnex.tests._utils_spmd import (
+    _generate_statistic_data,
+    _get_local_tensor,
+    _mpi_libs_and_gpu_available,
+)
+
+
+@pytest.mark.skipif(
+    not _mpi_libs_and_gpu_available,
+    reason="GPU device and MPI libs required for test",
+)
+@pytest.mark.parametrize(
+    "dataframe,queue",
+    get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu"),
+)
+@pytest.mark.parametrize("whiten", [True, False])
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+@pytest.mark.mpi
+def test_incremental_pca_fit_spmd_gold(dataframe, queue, whiten, dtype):
+    # Import spmd and non-SPMD algo
+    from sklearnex.preview.decomposition import IncrementalPCA
+    from sklearnex.spmd.decomposition import IncrementalPCA as IncrementalPCA_SPMD
+
+    # Create gold data and process into dpt
+    X = np.array(
+        [
+            [0.0, 0.0],
+            [1.0, 2.0],
+            [2.0, 4.0],
+            [3.0, 8.0],
+            [4.0, 16.0],
+            [5.0, 32.0],
+            [6.0, 64.0],
+            [7.0, 128.0],
+        ]
+    ).astype(dtype=dtype)
+    dpt_X = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+    local_X = _get_local_tensor(X)
+    local_dpt_X = _convert_to_dataframe(local_X, sycl_queue=queue, target_df=dataframe)
+
+    incpca_spmd = IncrementalPCA_SPMD(whiten=whiten)
+    incpca = IncrementalPCA(whiten=whiten)
+
+    incpca_spmd.fit(local_dpt_X)
+    incpca.fit(dpt_X)
+
+    assert_allclose(incpca.n_components_, incpca_spmd.n_components_)
+    assert_allclose(incpca.components_, incpca_spmd.components_)
+    assert_allclose(incpca.singular_values_, incpca_spmd.singular_values_)
+    assert_allclose(incpca.mean_, incpca_spmd.mean_)
+    assert_allclose(incpca.var_, incpca_spmd.var_)
+    assert_allclose(incpca.explained_variance_, incpca_spmd.explained_variance_)
+    assert_allclose(
+        incpca.explained_variance_ratio_, incpca_spmd.explained_variance_ratio_
+    )
```

Comment:
```suggestion
    for attr in incpca:
        if attr.endswith("_") and not attr.startswith("_"):
            assert_allclose(getattr(incpca, attr), getattr(incpa_spmd, attr), err_msg=attr)
```

Reply:
i'm not sure that it is better. as for me, current version is more stable, more readable and this is exactly the properties which the test should have

Reply:
so then why was this implemented? https://github.com/intel/scikit-learn-intelex/pull/1961#discussion_r1732094400

Reply:
because attribute list is variable there depending on provided result options, so there was no other way to implement this checks

Reply:
The attribute list is hardcoded based off of this, and is 'falsely' variable. https://github.com/intel/scikit-learn-intelex/blob/main/onedal/basic_statistics/tests/test_basic_statistics.py#L71 . I find your argument here to be invalid: https://github.com/intel/scikit-learn-intelex/pull/1979#discussion_r1742479955  You should have applied the same reasoning in that implementation.  I would change it here to be more stable, consistent, and make it clear in the error message as to what is failing.

Reply:
Also in this link, the options_and_tests is not parametrized and the bounds available in element[2] are not used either.  Your code is a direct corollary to this circumstance (https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/spmd/basic_statistics/tests/test_incremental_basic_statistics_spmd.py#L90), and I would just like consistency in your code.

Reply:
> I would change it here to be more stable, consistent, and make it clear in the error message as to what is failing.

it is not more stable, since other attributes could be added which satisfy your condition and it could lead to fail. regarding error message, it contains the line assert is failed, so currently it is as clear as in your suggestion.

> The attribute list is hardcoded

it is not true, there are hardcoded all possible result options. but fitted estimator has only the options provided in result_options.

Reply:
I highly recommend you to read the sklearn estimator development documentation, https://scikit-learn.org/stable/developers/develop.html#estimated-attributes .  Trailing underscores are fitted attributes, which would also likely need to be tested.  In that case, your approach would increase the maintenance burden, or cause incomplete coverage as the newly added attribute described in your argument would not be covered by your test any longer.

By transitive property, it is hardcoded because of the options in result_options is hardcoded. That is what I was getting at.

What if we just hardcoded the list of estimated attributes as a global in the testing file, would you switch it then?

Reply:
yes, this option is better, i'll do